### PR TITLE
Remove Bazel

### LINKED ISSUE #88553: Please remove bazel based build infrastructure
- it's tough to maintain 2 things in parallel and keep them in sync
- with the switch to go modules/cache, most of the reasons why bazel was used (speed up compilations) are gone
- gets in the way in multiple ways (example, update of golang versions)
- etc...

### PR DESCRIPTION
**Pull request by BenTheElder**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Removes the bazel build system, reducing Kubernetes to maintaining one build system for this repo per: https://github.com/kubernetes/enhancements/issues/2420

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/88553

#### Special notes for your reviewer:

The third commit "`hack/update-bazel.sh`" is fully automated, and is an enourmous diff. I recommend reviewing this PR commit by commit, skipping that one. The first commit implements the automation to produce this commit. 

Explicitly holding this PR, we need to send out a widespread announcement including additional notice of this change and instructions for how contributors should handle this. We need to wait before merging this (will send that tomorrow) given the scope of the change.

This PR also will not merge due to `pull-kubernetes-bazel-build` and `pull-kubernetes-bazel-test` which still running on all PRs as `Required` for now, which will fail.

/hold

All other known technical blockers for this change are resolved. See https://github.com/kubernetes/enhancements/issues/2420 for more details.

#### Does this PR introduce a user-facing change?

Technically this is not user facing, as our releases have always shipped with make, and is not under our deprecation policies as an API change would be. However those building from source should be aware of this change.

Interested downstream users can still build with bazel themselves if they wish, but we will not be carrying the patch load for this upstream going forward, instead focusing on the single system we use to ship releases.

```release-note
Remove support for building Kubernetes with bazel.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2420
```

/sig testing
/sig release
(this is going to tag all the SIGs anyhow, since it touches nearly all directories, we should probably remove those?)

### CODE REVIEW

File: hack/update-bazel.sh
Comment (BenTheElder) [inline hack/update-bazel.sh:22]: I'll be writing a more detailed notice but the TLDR on why this change is so authors with in-flight PRs containing bazel related diffs (e.g. because they added a go source file) can just do:
- rebase to upstream after this merges
- run `hack/update-bazel.sh` (now _very_ fast) during the rebase to drop all bazel related generated files
- carry on with their PR as usual

**Comment by BenTheElder** [comment]
Changes in my last push:
- added `${KUBE_ROOT}/build/root/BUILD.root` to auto-removed files logic in first commit
- regenerated the `hack/update-bazel.sh` commit after this
  - `find . -name 'BUILD.*'` now reports nothing.
- added an additional commit to manually drop `OWNERS`, `README.md` from directories that formerly contained bazel files and now contain nothing else, cleaning up these dangling directories.

**Comment by BenTheElder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99561/pull-kubernetes-verify/1366165916764606464
one verify failure fixed by the latest commit. (`hack/verify-openapi-spec.sh` was copying `BUILD` specifically into the directory to generate in)

The `pull-kubernetes-conformance-image-test` failures are not related to this PR, building the image is flaking on pulling the base image, which is an unrelated issue. It's also not a required test.

**Comment by liggitt** [comment]
reviewed locally since github won't show the full diff

/approve

lgtm once the bazel CI jobs are dropped from master

**Comment by liggitt** [comment]
pull-kubernetes-verify shows running the conformance verifications:
```
verify: conformance-requirements
verify: conformance-yaml
```

**Comment by derekwaynecarr** [comment]
this should get a medal for "removed most lines in a single PR"

/approve (for cri-api)

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
/remove-priority important-soon
/priority critical-urgent
/kind flake
since technically bazel's build flakes on fetching dependencies at times

**Comment by serathius** [comment]
Hey @BenTheElder, Thanks for your hard work on this. I have one question, SIG Instrumentation depended on bazel for forbidding imports of particular libraries (https://github.com/kubernetes/kubernetes/blob/a1f2787897a81c375a4de28482b5eb649851800a/build/visible_to/BUILD#L428), thus ensuring that contributors will use Kubernetes alternatives (context https://github.com/kubernetes/kubernetes/issues/89267). What is the current replacement for this baze feature?

**Comment by BenTheElder** [comment]
Good question, there's not a _direct_ analog, but we have some related alternatives available:
- go's built-in `internal` concept (you probably knew that, it's not quite as flexible / requires codepath organization to align)
- "import boss" https://github.com/kubernetes/kubernetes/blob/8306eeab75737d431068fd0551770abf6da3c1cf/hack/verify-import-boss.sh (our own custom go tool wrapped by this script run in `pull-kubernetes-verify)

For this particular case of "restrict imports of a vendored library" internal does not apply but import boss should work. I can help sort this out.

**Comment by BenTheElder** [comment]
... And apologies for the lapse in coverage in the meantime. Clearly an oversight on my part. I'll look at equivilant rules shortly. Import-Boss is what we typically use to restrict package visibility within the project, but typically for packages within our own sources. It may need some adapation to cover vendor / external packages.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/99876 to track visibility rules -> `make verify`

**Comment by duchuanLX** [comment]
> #99876 to track visibility rules -> `make verify`
I want to know whether the follow-up community build completely cancels bazel, only uses make, and make is a stable build tool for subsequent versions.

**Comment by BenTheElder** [comment]
Hi all: please direct general discussion to the KEP if possible for better visibility. This PR is only one peice of implementing it. https://github.com/kubernetes/enhancements/issues/2420#issuecomment-791024902

The KEP answers your questions I think.
https://github.com/kubernetes/enhancements/issues/2420
https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/2420-reducing-kubernetes-build-maintenance/

Bazel is not in use in this repo from 1.21 / this PR forward. Bazel was not used for releases, only CI and optionally for development.

**Pull request merged**
